client/rest/src/main/java/org/opensearch/client/RestClient.java

```diff
@@ -832,6 +832,7 @@ private class InternalRequest {
         InternalRequest(Request request) {
             this.request = request;
             Map<String, String> params = new HashMap<>(request.getParameters());
+            params.putAll(request.getOptions().getQueryParams());
```

Comment:
`request.getParameters()` should do the job I believe

Reply:
@reta I'm not sure I agree.  Per my understanding and the implementation I have written, this is the only place (apart from the `toString()`, of course!) where the new `parameters` map on the `RequestOptions` object is retrieved for adding to the actual `Request`.

Or have I misunderstood something?

Reply:
@oliverlockwood here is my reasoning: as the developer, I would expect `request.getParameters()` to give me all parameters (== query parameters) needed for this particular request. Now, the `request` has not only own `parameters` but the ones which come from `RequestOptions` as well. It brings two questions:
 - at what circumstances the `request.getParameters()` should be used but `request.getOptions().getParameters()` should not?
 - how exactly the consumer of this API should be aware that there are also parameters on  `request.getOptions()` to be taken into account and not only  `request.getParameters()`?

Does it explain my concerns? Thank you.

Reply:
OK, I think I'm with you now.  Essentially, I believe you're suggesting to change the implementation of the `Request.getParameters()`  method which is here: https://github.com/opensearch-project/OpenSearch/blob/main/client/rest/src/main/java/org/opensearch/client/Request.java#L112:L114.

Pasting the code including comment:
```java
    /**
     * Query string parameters. The returned map is an unmodifiable view of the
     * map in the request so calls to {@link #addParameter(String, String)}
     * will change it.
     */
    public Map<String, String> getParameters() {
        return unmodifiableMap(parameters);
    }
```

The way forward is not 100% obvious to me because, as per the implementation & Javadoc, it seems quite possible that there may be usages of `Request` which rely on being able to dynamically mutate the `parameters` map, even after a consumer has called `getParameters()`. I've come up with a couple of options, as follows.

**Option 1** something like
```java
    public Map<String, String> getParameters() {
        Map<String, String> combinedParameters = new HashMap<>(parameters);
        parameters.putAll(options.getParameters());
        return unmodifiableMap(combinedParameters);
    }
```
This combines the option parameters when requested, but loses the dynamic-mutation capability.

**Option 2**
Change `Request.setOptions()` method to add all the parameters in the `RequestOptions` into the `Request`.
This retains the dynamic-mutation capability at `Request` level, but if parameters are subsequently added to the `RequestOptions`, before the `Request` is sent, they will not be respected.

Of the two, I suspect Option 2 would be preferable, but I'd also be very happy to hear of any better ideas you may have, as doubtless you are more familiar with the codebase and usecases.

Please let me know how you'd like me to proceed.

Reply:
Thanks @oliverlockwood , the:

```
   public Map<String, String> getParameters() {
        if (options.getParameters().isEmpty() {
          return unmodifiableMap(parameters);
        } else {
          Map<String, String> combinedParameters = new HashMap<>(parameters);
          parameters.putAll(options.getParameters());
          return unmodifiableMap(combinedParameters);
        }
    }
```

is perfectly fine. I think we should also document on `RequestOptions` that the `parameters` will take precedence over any request specific parameters. Thank you.

Reply:
@reta it's now changed per your feedback.  Please re-review when you have the chance  Many thanks.